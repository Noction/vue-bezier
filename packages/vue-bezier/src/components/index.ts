// Exporting components from Collapse directory
export { default as CollapseTransition } from './Collapse/CollapseTransition.vue';

// Exporting components from Fade directory
export { default as FadeTransition } from './Fade/FadeTransition.vue';

// Exporting components from Scale directory
export { default as ScaleTransition } from './Scale/ScaleTransition.vue';

// Exporting components from Slide directory
export { default as SlideXLeftTransition } from './Slide/SlideXLeftTransition.vue';
export { default as SlideXRightTransition } from './Slide/SlideXRightTransition.vue';
export { default as SlideYDownTransition } from './Slide/SlideYDownTransition.vue';
export { default as SlideYUpTransition } from './Slide/SlideYUpTransition.vue';

// Exporting components from Zoom directory
export { default as ZoomCenterTransition } from './Zoom/ZoomCenterTransition.vue';
export { default as ZoomUpTransition } from './Zoom/ZoomUpTransition.vue';
export { default as ZoomXTransition } from './Zoom/ZoomXTransition.vue';
export { default as ZoomYTransition } from './Zoom/ZoomYTransition.vue';
