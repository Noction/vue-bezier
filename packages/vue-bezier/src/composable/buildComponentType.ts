import { Component, Transition, TransitionGroup } from 'vue'

export default (props): Component =>
  props.group ? TransitionGroup : Transition
