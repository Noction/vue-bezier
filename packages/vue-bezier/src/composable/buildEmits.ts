export default () => [
  'after-enter',
  'after-leave',
  'before-enter',
  'before-leave',
  'leave'
]
