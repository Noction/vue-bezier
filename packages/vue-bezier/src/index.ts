import PrivateCollapseTransition from './components/Collapse/CollapseTransition.vue'

import PrivateFadeTransition from './components/Fade/FadeTransition.vue'

import PrivateScaleTransition from './components/Scale/ScaleTransition.vue'

import PrivateSlideXLeftTransition from './components/Slide/SlideXLeftTransition.vue'
import PrivateSlideXRightTransition from './components/Slide/SlideXRightTransition.vue'
import PrivateSlideYDownTransition from './components/Slide/SlideYDownTransition.vue'
import PrivateSlideYUpTransition from './components/Slide/SlideYUpTransition.vue'

import PrivateZoomCenterTransition from './components/Zoom/ZoomCenterTransition.vue'
import PrivateZoomUpTransition from './components/Zoom/ZoomUpTransition.vue'
import PrivateZoomXTransition from './components/Zoom/ZoomXTransition.vue'
import PrivateZoomYTransition from './components/Zoom/ZoomYTransition.vue'

const components = {
  CollapseTransition: PrivateCollapseTransition,
  FadeTransition: PrivateFadeTransition,
  ScaleTransition: PrivateScaleTransition,
  SlideXLeftTransition: PrivateSlideXLeftTransition,
  SlideXRightTransition: PrivateSlideXRightTransition,
  SlideYDownTransition: PrivateSlideYDownTransition,
  SlideYUpTransition: PrivateSlideYUpTransition,
  ZoomCenterTransition: PrivateZoomCenterTransition,
  ZoomUpTransition: PrivateZoomUpTransition,
  ZoomXTransition: PrivateZoomXTransition,
  ZoomYTransition: PrivateZoomYTransition
}

export {
  PrivateCollapseTransition as CollapseTransition,
  PrivateFadeTransition as FadeTransition,
  PrivateScaleTransition as ScaleTransition,
  PrivateSlideXLeftTransition as SlideXLeftTransition,
  PrivateSlideXRightTransition as SlideXRightTransition,
  PrivateSlideYDownTransition as SlideYDownTransition,
  PrivateSlideYUpTransition as SlideYUpTransition,
  PrivateZoomCenterTransition as ZoomCenterTransition,
  PrivateZoomUpTransition as ZoomUpTransition,
  PrivateZoomXTransition as ZoomXTransition,
  PrivateZoomYTransition as ZoomYTransition
}

export function install (app) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (app.$_v3SFCTransitionsInstalled) return
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  app.$_v3SFCTransitionsInstalled = true

  Object
    .entries(components)
    .forEach(([key, component]) =>
      app.component(key, component)
    )
}

export default {
  install
}
