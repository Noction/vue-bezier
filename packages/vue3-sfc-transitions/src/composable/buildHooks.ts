import { BaseTransitionProps } from 'vue'

const beforeEnter = (props, el: HTMLElement) => {
  const enterDuration = props.duration?.enter ?? props.duration ?? 0
  const enterDelay = props.delay?.enter ?? props.delay ?? 0

  el.style.animationDuration = `${enterDuration}ms`
  el.style.animationDelay = `${enterDelay}ms`

  setStyles(props, el)
}

const beforeLeave = (props, el: HTMLElement) => {
  const leaveDuration = props.duration?.leave ?? props.duration ?? 0
  const leaveDelay = props.delay?.leave ?? props.delay ?? 0

  el.style.animationDuration = `${leaveDuration}ms`
  el.style.animationDelay = `${leaveDelay}ms`

  setStyles(props, el)
}

const cleanUpStyles = (props, el: HTMLElement) => {
  Object
    .keys(props.styles)
    .forEach(key => {
      const styleValue = props.styles[key]

      if (styleValue) el.style[key] = ''
    })

  el.style.animationDuration = ''
  el.style.animationDelay = ''
}

const leave = (props, el: HTMLElement, done: () => void) => {
  setAbsolutePosition(props, el)

  const leaveDuration = props.duration?.leave ?? props.duration ?? 0
  const leaveDelay = props.delay?.leave ?? props.delay ?? 0

  setTimeout(done, leaveDuration + leaveDelay)
}

const setStyles = (props, el: HTMLElement) => {
  setTransformOrigin(props, el)

  Object
    .keys(props.styles)
    .forEach(key => {
      const styleValue = props.styles[key]

      if (styleValue) el.style[key] = styleValue
    })
}

const setAbsolutePosition = (props, el: HTMLElement) => {
  if (props.group) el.style.position = 'absolute'
}

const setTransformOrigin = (props, el: HTMLElement) => {
  if (props.origin) el.style.transformOrigin = props.origin
}

export default (props, emit): BaseTransitionProps => ({
  onAfterEnter: (el: HTMLElement) => {
    cleanUpStyles(props, el)
    emit('after-enter', el)
  },
  onAfterLeave: (el: HTMLElement) => {
    cleanUpStyles(props, el)
    emit('after-leave', el)
  },
  onBeforeEnter: (el: HTMLElement) => {
    beforeEnter(props, el)
    emit('before-enter', el)
  },
  onBeforeLeave: (el: HTMLElement) => {
    beforeLeave(props, el)
    emit('before-leave', el)
  },
  onLeave: (el: HTMLElement, done: () => void) => {
    leave(props, el, done)
    emit('leave', el, done)
  }
})

export {
  setAbsolutePosition,
  setStyles
}
