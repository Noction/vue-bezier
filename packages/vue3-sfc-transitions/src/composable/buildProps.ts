const defaultProps = {
  /**
   * Transition delay. Number for specifying the same delay for enter/leave transitions
   * Object style {enter: 300, leave: 300} for specifying explicit durations for enter/leave
   */
  delay: {
    default: 0,
    type: [Number, Object]
  },
  /**
   * Transition duration. Number for specifying the same duration for enter/leave transitions
   * Object style {enter: 300, leave: 300} for specifying explicit durations for enter/leave
   */
  duration: {
    default: 300,
    type: [Number, Object]
  },
  /**
   * Whether the component should be a `transition-group` component.
   */
  group: Boolean,
  /**
   *  Transform origin property https://tympanus.net/codrops/css_reference/transform-origin/.
   *  Can be specified with styles as well but it's shorter with this prop
   */
  origin: {
    default: '',
    type: String
  },
  /**
   * Element styles that are applied during transition. These styles are applied on @beforeEnter and @beforeLeave hooks
   */
  styles: {
    default: () => ({
      animationFillMode: 'both',
      animationTimingFunction: 'ease-out'
    }),
    type: Object
  },
  /**
   * Transition tag, in case the component is a `transition-group`
   */
  tag: {
    default: 'span',
    type: String
  }
}

export default (customProps?: Record<string, unknown>) => ({
  ...defaultProps,
  ...(customProps || {})
})
