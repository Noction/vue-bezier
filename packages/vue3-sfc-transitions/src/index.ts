import PrivateCollapseTransition from './components/Collapse/CollapseTransition.vue'

import PrivateFadeTransition from './components/Fade/FadeTransition.vue'

import PrivateScaleTransition from './components/Scale/ScaleTransition.vue'

import PrivateSlideXLeftTransition from './components/Slide/SlideXLeftTransition.vue'
import PrivateSlideXRightTransition from './components/Slide/SlideXRightTransition.vue'
import PrivateSlideYDownTransition from './components/Slide/SlideYDownTransition.vue'
import PrivateSlideYUpTransition from './components/Slide/SlideYUpTransition.vue'

import PrivateZoomCenterTransition from './components/Zoom/ZoomCenterTransition.vue'
import PrivateZoomUpTransition from './components/Zoom/ZoomUpTransition.vue'
import PrivateZoomXTransition from './components/Zoom/ZoomXTransition.vue'
import PrivateZoomYTransition from './components/Zoom/ZoomYTransition.vue'

interface Options {
  casing?: 'PascalCase' | 'kebab-case'
  components?: Array<
    | 'FadeTransition'
    | 'ZoomCenterTransition'
    | 'ZoomXTransition'
    | 'ZoomYTransition'
    | 'ZoomUpTransition'
    | 'SlideYUpTransition'
    | 'SlideYDownTransition'
    | 'SlideXLeftTransition'
    | 'SlideXRightTransition'
    | 'ScaleTransition'
    | 'CollapseTransition'
  >
}

const components = {
  CollapseTransition: PrivateCollapseTransition,
  FadeTransition: PrivateFadeTransition,
  ScaleTransition: PrivateScaleTransition,
  SlideXLeftTransition: PrivateSlideXLeftTransition,
  SlideXRightTransition: PrivateSlideXRightTransition,
  SlideYDownTransition: PrivateSlideYDownTransition,
  SlideYUpTransition: PrivateSlideYUpTransition,
  ZoomCenterTransition: PrivateZoomCenterTransition,
  ZoomUpTransition: PrivateZoomUpTransition,
  ZoomXTransition: PrivateZoomXTransition,
  ZoomYTransition: PrivateZoomYTransition
}

export {
  PrivateCollapseTransition as CollapseTransition,
  PrivateFadeTransition as FadeTransition,
  PrivateScaleTransition as ScaleTransition,
  PrivateSlideXLeftTransition as SlideXLeftTransition,
  PrivateSlideXRightTransition as SlideXRightTransition,
  PrivateSlideYDownTransition as SlideYDownTransition,
  PrivateSlideYUpTransition as SlideYUpTransition,
  PrivateZoomCenterTransition as ZoomCenterTransition,
  PrivateZoomUpTransition as ZoomUpTransition,
  PrivateZoomXTransition as ZoomXTransition,
  PrivateZoomYTransition as ZoomYTransition
}

export function install (app, options?: Options) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (app.$_v3SFCTransitionsInstalled) return
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  app.$_v3SFCTransitionsInstalled = true

  const componentNameCasing = {
    PascalCase: str => str,
    'kebab-case': str =>
      `${str.slice(0, 1)}${str
        .slice(1)
        .replace(/[A-Z]/g, '-$&')}`.toLocaleLowerCase()
  }

  const nameFormat = componentNameCasing[options?.casing ?? 'kebab-case']

  if (options && options?.components) {
    options.components.forEach((name: string) =>
      app.component(nameFormat(name), components[name])
    )
  } else {
    Object.entries(components).forEach(([key, component]) =>
      app.component(nameFormat(key), component)
    )
  }
}

export default {
  install
}
