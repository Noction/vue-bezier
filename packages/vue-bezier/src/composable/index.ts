import buildComponentType from './buildComponentType'
import buildEmits from './buildEmits'
import buildHooks from './buildHooks'
import buildProps from './buildProps'
import buildTag from './buildTag'

export { buildComponentType, buildEmits, buildHooks, buildProps, buildTag }
