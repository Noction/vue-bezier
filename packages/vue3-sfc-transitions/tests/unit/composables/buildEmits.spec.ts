import { buildEmits } from '../../../src/composable'

describe('composable / buildEmits', () => {
  it('The returned list of emit should match the snapshot',  () => {
    expect(buildEmits()).toMatchSnapshot()
  })
})
