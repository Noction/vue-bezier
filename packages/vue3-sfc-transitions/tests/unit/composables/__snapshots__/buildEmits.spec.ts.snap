// Vitest Snapshot v1

exports[`composable / buildEmits > The returned list of emit should match the snapshot 1`] = `
[
  "after-enter",
  "after-leave",
  "before-enter",
  "before-leave",
  "leave",
]
`;
